### PR TITLE
eliminate blocking in initialization, #6

### PR DESCRIPTION
**Pull request by patriknw**
* and several other places
* The remaining blocking is in the RowIterator that is used
  for loading snapshot metadata and the RowIterator that is
  used when deleting events (ReadLowestSequenceNr). That
  blocking is now contained by a separate dispatcher and
  we can improve that later.
* I have verified (with CassandraJournalPerfSpec) that
  there is no noticeable performance regression.

### CODE REVIEW

**Comment by patriknw** [comment]
Refs #6

**Comment by patriknw** [comment]
Created issue https://github.com/akka/akka-persistence-cassandra/issues/75 for the RowIterator

**Comment by patriknw** [comment]
If you have time it would great if you can review this refactoring @zapletal-martin to have a second pair of eyes looking for mistakes in the refactoring.

File: src/main/resources/reference.conf [hunk 1]
```diff
@@ -187,21 +187,14 @@ cassandra-journal {
...
-    fork-join-executor {
-      parallelism-min = 2
-      parallelism-max = 8
-    }
-  }
+  # Dispatcher for the plugin actor.
+  plugin-dispatcher = "cassandra-plugin-default-dispatcher"
+  
+  # Dispatcher for potentially blocking tasks.
+  blocking-dispatcher = "cassandra-plugin-blocking-dispatcher"
```
Comment (ktoso) [inline src/main/resources/reference.conf]: :+1:

File: src/main/scala/akka/persistence/cassandra/CassandraSession.scala [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+/**
+ * INTERNAL API
+ */
+private[cassandra] final class CassandraSession(
+  system: ActorSystem, settings: CassandraPluginConfig, executionContext: ExecutionContext,
+  log:             LoggingAdapter,
+  metricsCategory: String,
+  init:            Session => Future[_]
+) {
```
Comment (ktoso) [inline src/main/scala/akka/persistence/cassandra/CassandraSession.scala]: new scalariform ay?

File: src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala [hunk 3]
```diff
@@ -136,23 +99,22 @@ class CassandraJournal(cfg: Config) extends AsyncWriteJournal with CassandraReco
...
-      }
+      // try initialize early, to be prepared for first real request
+      preparedWriteMessage
+      preparedDeleteMessages
+      preparedSelectMessages
+      preparedCheckInUse
+      preparedWriteInUse
+      preparedSelectHighestSequenceNr
+      preparedSelectDeletedTo
+      preparedInsertDeletedTo
```
Comment (ktoso) [inline src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala]: :+1:

File: src/main/scala/akka/persistence/cassandra/journal/CassandraStatements.scala [hunk 4]
```diff
@@ -143,22 +145,71 @@ trait CassandraStatements {
...
-        session.execute(createKeyspace)
-      session.execute(createTable)
-      session.execute(createMetatdataTable)
-      session.execute(createConfigTable)
-      for (tagId <- 1 to maxTagId)
-        session.execute(createEventsByTagMaterializedView(tagId))
+  def executeCreateKeyspaceAndTables(session: Session, keyspaceAutoCreate: Boolean,
+                                     maxTagId: Int)(implicit ec: ExecutionContext): Future[Done] = {
+    import akka.persistence.cassandra.listenableFutureToFuture
+    CassandraSession.createKeyspaceAndTablesSemaphore.acquire()
```
Comment (ktoso) [inline src/main/scala/akka/persistence/cassandra/journal/CassandraStatements.scala]: We'll want to replace those with future sequencing eventually?

**Comment by ktoso** [comment]
Looks good overall, but I didn't look very deep..

File: src/main/scala/akka/persistence/cassandra/journal/CassandraStatements.scala (hunk 4 above)
Comment (patriknw) [inline src/main/scala/akka/persistence/cassandra/journal/CassandraStatements.scala]: I can try that, might work with some CASing

**Comment by patriknw** [comment]
Rebased and rewrote the serialized execution of keyspace and table creation. 
@ktoso please sanity check 9d83d88e6746a4bc0b87139e7d87a37a09b4d119

File: src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala [hunk 5]
```diff
@@ -300,59 +269,67 @@ class CassandraJournal(cfg: Config) extends AsyncWriteJournal with CassandraReco
...
+              def asyncDeleteMessages(partitionNr: Long, messageIds: Seq[MessageId]): Future[Unit] = {
+                val boundStatements = messageIds.map(mid =>
+                  preparedDeleteMessages.map(_.bind(mid.persistenceId, partitionNr: JLong, mid.sequenceNr: JLong)))
+                Future.sequence(boundStatements).flatMap { stmts =>
+                  executeBatch(batch => stmts.foreach(batch.add), deleteRetryPolicy)
+                }
+              }
+
+              partitionInfos.map(future => future.flatMap(pi => {
+                Future.sequence((pi.minSequenceNr to pi.maxSequenceNr).grouped(config.maxMessageBatchSize).map { group =>
```
Comment (patriknw) [inline src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala]: @ktoso looks like this is wrong. `pi.maxSequenceNr` should be replaced with `min(toSeqNr, pi.maxSequenceNr`), I think

Comment (ktoso) [inline src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala]: Yeah, exactly :+1:

File: src/main/scala/akka/persistence/cassandra/CassandraSession.scala [hunk 6]
```diff
@@ -165,5 +165,21 @@ private[cassandra] final class CassandraSession(
  * INTERNAL API
  */
 private[cassandra] final object CassandraSession {
-  val createKeyspaceAndTablesSemaphore = new Semaphore(1)
+  private val serializedExecutionProgress = new AtomicReference[Future[Done]](Future.successful(Done))
+
+  def serializedExecution(recur: () => Future[Done], exec: () => Future[Done])(implicit ec: ExecutionContext): Future[Done] = {
```
Comment (ktoso) [inline src/main/scala/akka/persistence/cassandra/CassandraSession.scala]: takes some wrapping your head around but looks valid, LGTM

**Pull request merged**